File: src/Apps/NetPad.Apps.App/App/src/core/@common/events/subscription-token.ts
import {IDisposable} from "@common";

export class SubscriptionToken implements IDisposable {
    private disposeAction: (() => void) | null;
    private isDisposed = false;

    constructor(disposeAction: () => void) {
        this.disposeAction = disposeAction;
    }

    public dispose(): void {
        if (this.isDisposed) return;

        try{
            if (this.disposeAction) {
                this.disposeAction();
            }
        } finally {
            this.isDisposed = true;
            this.disposeAction = null;
        }
    }
}

File: src/External/OmniSharp.NET/Events/SubscriptionToken.cs
using System;

namespace OmniSharp.Events
{
    public sealed class SubscriptionToken : IDisposable
    {
        private readonly Action _unsubscribe;

        internal SubscriptionToken(Action unsubscribe)
        {
            _unsubscribe = unsubscribe;
        }

        /// <summary>
        /// Unsubscribes the token.
        /// </summary>
        public void Dispose()
        {
            _unsubscribe();
        }
    }
}
